File: gnap-compliance/src/validation/grant_response.rs
/// Grant response validation — RFC 9635 Section 3
use crate::types::{AccessTokenResponseField, ComplianceError, GrantResponse};

/// Validate a grant response for RFC 9635 compliance.
pub fn validate_grant_response(resp: &GrantResponse) -> Result<(), Vec<ComplianceError>> {
    let mut errors = Vec::new();

    // A grant response must contain at least one of: continue, access_token, interact, subject, error
    // RFC 9635 Section 3
    if resp.continue_.is_none()
        && resp.access_token.is_none()
        && resp.interact.is_none()
        && resp.subject.is_none()
        && resp.instance_id.is_none()
        && resp.error.is_none()
    {
        errors.push(ComplianceError::Validation(
            "Grant response must contain at least one field (Section 3)".to_string(),
        ));
    }

    // Validate continue if present
    if let Some(cont) = &resp.continue_ {
        if cont.access_token.value.is_empty() {
            errors.push(ComplianceError::Validation(
                "continue.access_token.value must not be empty (Section 3.1)".to_string(),
            ));
        }
        if cont.uri.is_empty() {
            errors.push(ComplianceError::Validation(
                "continue.uri must not be empty (Section 3.1)".to_string(),
            ));
        }
    }

    // Validate access_token if present
    if let Some(token_field) = &resp.access_token {
        validate_access_token_response(token_field, &mut errors);
    }

    // RFC 9635 Section 3.6: error MUST NOT be combined with access_token, interact, subject, instance_id
    if resp.error.is_some()
        && (resp.access_token.is_some()
            || resp.interact.is_some()
            || resp.subject.is_some()
            || resp.instance_id.is_some())
    {
        errors.push(ComplianceError::Validation(
            "When error is present, only continue may also be present (Section 3.6)".to_string(),
        ));
    }

    if errors.is_empty() {
        Ok(())
    } else {
        Err(errors)
    }
}

fn validate_single_token(
    token: &crate::types::AccessToken,
    prefix: &str,
    errors: &mut Vec<ComplianceError>,
) {
    if token.value.is_empty() {
        errors.push(ComplianceError::Validation(format!(
            "{prefix}.value must not be empty (Section 3.2)"
        )));
    }
    if token.access.is_empty() {
        errors.push(ComplianceError::Validation(format!(
            "{prefix}.access must not be empty (Section 3.2.1)"
        )));
    }
    // Section 3.2.1: bearer flag and key field are mutually exclusive.
    if let Some(flags) = &token.flags {
        if flags.contains(&"bearer".to_string()) && token.key.is_some() {
            errors.push(ComplianceError::Validation(format!(
                "{prefix}: bearer flag and key must not both be present (Section 3.2.1)"
            )));
        }
        // Section 3.2.1: flag values MUST NOT be duplicated.
        let mut seen = std::collections::HashSet::new();
        for flag in flags {
            if !seen.insert(flag) {
                errors.push(ComplianceError::Validation(format!(
                    "{prefix}: duplicate flag \"{flag}\" (Section 3.2.1)"
                )));
            }
        }
    }
    // Section 3.2.1: management token value MUST differ from managed token value.
    if let Some(manage) = &token.manage {
        if manage.access_token.value == token.value {
            errors.push(ComplianceError::Validation(format!(
                "{prefix}: manage.access_token.value must differ from token value (Section 3.2.1)"
            )));
        }
    }
}

fn validate_access_token_response(
    field: &AccessTokenResponseField,
    errors: &mut Vec<ComplianceError>,
) {
    match field {
        AccessTokenResponseField::Single(token) => {
            validate_single_token(token, "access_token", errors);
        }
        AccessTokenResponseField::Multiple(tokens) => {
            if tokens.is_empty() {
                errors.push(ComplianceError::Validation(
                    "access_token array must not be empty (Section 3.2)".to_string(),
                ));
            }
            for (i, token) in tokens.iter().enumerate() {
                validate_single_token(token, &format!("access_token[{i}]"), errors);
                if token.label.is_none() {
                    errors.push(ComplianceError::Validation(format!(
                        "access_token[{i}].label is required for multi-token responses (Section 3.2)"
                    )));
                }
            }
            if let Some(e) = super::helpers::check_labels_unique(
                tokens.iter().map(|t| t.label.as_deref()),
                "Section 3.2",
            ) {
                errors.push(e);
            }
        }
    }
}

File: gnap-client/src/validation/grant-response.ts
/** Grant response validation -- RFC 9635 Section 3 */

import type { GrantResponse } from "../types/grant.ts";
import type { AccessTokenResponseField, AccessToken } from "../types/token.ts";
import { isMultipleTokenResponse } from "../types/token.ts";

function checkLabelsUnique(
  labels: (string | undefined)[],
  section: string,
): string | null {
  const seen = new Set<string>();
  for (const label of labels) {
    if (label !== undefined) {
      if (seen.has(label)) {
        return `Duplicate label "${label}" (${section})`;
      }
      seen.add(label);
    }
  }
  return null;
}

function validateSingleToken(
  token: AccessToken,
  prefix: string,
  errors: string[],
): void {
  if (token.value === "") {
    errors.push(`${prefix}.value must not be empty (Section 3.2)`);
  }
  if (token.access.length === 0) {
    errors.push(`${prefix}.access must not be empty (Section 3.2.1)`);
  }
  if (token.flags) {
    if (token.flags.includes("bearer") && token.key !== undefined) {
      errors.push(
        `${prefix}: bearer flag and key must not both be present (Section 3.2.1)`,
      );
    }
    const seen = new Set<string>();
    for (const flag of token.flags) {
      if (seen.has(flag)) {
        errors.push(`${prefix}: duplicate flag "${flag}" (Section 3.2.1)`);
      }
      seen.add(flag);
    }
  }
  if (token.manage && token.manage.access_token.value === token.value) {
    errors.push(
      `${prefix}: manage.access_token.value must differ from token value (Section 3.2.1)`,
    );
  }
}

function validateAccessTokenResponse(
  field: AccessTokenResponseField,
  errors: string[],
): void {
  if (isMultipleTokenResponse(field)) {
    if (field.length === 0) {
      errors.push("access_token array must not be empty (Section 3.2)");
    }
    for (let i = 0; i < field.length; i++) {
      const token = field[i]!;
      validateSingleToken(token, `access_token[${i}]`, errors);
      if (token.label === undefined) {
        errors.push(
          `access_token[${i}].label is required for multi-token responses (Section 3.2)`,
        );
      }
    }
    const dup = checkLabelsUnique(
      field.map((t) => t.label),
      "Section 3.2",
    );
    if (dup) errors.push(dup);
  } else {
    validateSingleToken(field, "access_token", errors);
  }
}

/** Validate a grant response for RFC 9635 compliance. Returns an array of error messages. */
export function validateGrantResponse(resp: GrantResponse): string[] {
  const errors: string[] = [];

  if (
    !resp.continue &&
    !resp.access_token &&
    !resp.interact &&
    !resp.subject &&
    !resp.instance_id &&
    !resp.error
  ) {
    errors.push(
      "Grant response must contain at least one field (Section 3)",
    );
  }

  if (resp.continue) {
    if (resp.continue.access_token.value === "") {
      errors.push(
        "continue.access_token.value must not be empty (Section 3.1)",
      );
    }
    if (resp.continue.uri === "") {
      errors.push("continue.uri must not be empty (Section 3.1)");
    }
  }

  if (resp.access_token) {
    validateAccessTokenResponse(resp.access_token, errors);
  }

  if (
    resp.error !== undefined &&
    (resp.access_token !== undefined ||
      resp.interact !== undefined ||
      resp.subject !== undefined ||
      resp.instance_id !== undefined)
  ) {
    errors.push(
      "When error is present, only continue may also be present (Section 3.6)",
    );
  }

  return errors;
}
